prog/dftb+/lib_dftb/populations.F90

```diff
@@ -451,5 +452,85 @@ subroutine getChargePerShell(qOrb, orb, species, chargePerShell, qRef)
   end subroutine getChargePerShell
 
 
+  subroutine denseBlockMulliken(rhoSqr, overSqr, q0, iSquare, qq)
+
+    !> Square (lower triangular) spin polarized density matrix
+    real(dp), intent(in) :: rhoSqr(:,:,:)
+
+    !> Square (lower triangular) overlap matrix
+    real(dp), intent(in) :: overSqr(:,:)
+
+    !> reference charges
+    real(dp), intent(in) :: q0(:,:,:)
+
+    !> Atom positions in the row/column of square matrices
+    integer, intent(in) :: iSquare(:)
+
+    !> Mulliken block charges on output (mOrb, mOrb, nAtom, nSpin)
+    real(dp), intent(out) :: qq(:,:,:,:)
+
+    real(dp), allocatable :: tmpS(:,:)
+    real(dp), allocatable :: tmpD(:,:,:)
+    real(dp), allocatable :: tmp_mat(:,:)
+    real(dp), allocatable :: tmp_block(:,:)
+
+    integer :: nAOs, nat, nSpin, ii, jj, kk, iat, iS
+
+    nSpin = size(rhoSqr, dim=3)
+    nat = size(iSquare, dim=1) - 1
+    nAOs = size(rhoSqr, dim=1)
+
+    ALLOCATE(tmpS(nAOs,nAOs))
+    ALLOCATE(tmpD(nAOs,nAOs,nSpin))
+    ALLOCATE(tmp_mat(nAOs,nAOs))
+    ALLOCATE(tmp_block(nAOs,nAOs))
+
+    ! ... symmetrize overlap and density matrices
+    tmpS(:,:) = 0.0_dp
+    tmpS(:,:) = overSqr(:,:) + transpose(overSqr(:,:))
```

Comment:
On the right hand side of assignments, specifying (:,:) might lead to some compilers making a copy of this variable. You cannot avoid it for array slicing, but otherwise
tmpS(:,:) = overSqr + transpose(overSqr)
is better.